File: 1338-reduce-array-size-to-the-half/1338-reduce-array-size-to-the-half.go
func minSetSize(arr []int) int {
   c := make([]int, 100001)
	r, l := 0, len(arr)

	for _, i := range arr {
		c[i]++
	}

	sort.Slice(c, func(i, j int) bool {
		return c[i] > c[j]
	})

	for i := 0; i < len(c) && c[i] != 0; i++ {
		l -= c[i]
		r++
		if l <= len(arr)/2 {
			return r
		}
	}
	
	return len(arr)
}

File: 1338-reduce-array-size-to-the-half/1338-reduce-array-size-to-the-half.py
class Solution:
    def minSetSize(self, arr: List[int]) -> int:
        cnt = Counter(arr)
        frequencies = list(cnt.values())
        frequencies.sort()
        
        ans, removed, half = 0, 0, len(arr) // 2
        while removed < half:
            ans += 1
            removed += frequencies.pop()
        return ans